sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1339,20 +1396,38 @@ def task_worker(self):
                            api.tx_disable_channel(host_lanes, True)
                            self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_READY
                            continue
+                    # Configure the target output power if ZR module
+                        if api.is_coherent_module():
+                           tx_power = self.port_dict[lport]['tx_power']
+                           # Prevent configuring same tx power multiple times
+                           if 0 != tx_power and tx_power != api.get_tx_config_power():
+                              if 1 != self.configure_tx_output_power(api, lport, tx_power):
+                                 self.log_error("{} failed to configure Tx power = {}".format(lport, tx_power))
+                              else:
+                                 self.log_notice("{} Successfully configured Tx power = {}".format(lport, tx_power))
 
                         appl = self.get_cmis_application_desired(api, host_lanes, host_speed)
                         if appl < 1:
                             self.log_error("{}: no suitable app for the port".format(lport))
                             self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_FAILED
                             continue
 
-                        has_update = self.is_cmis_application_update_required(api, host_lanes, host_speed)
-                        if not has_update:
+                        need_update = self.is_cmis_application_update_required(api, host_lanes, host_speed)
+
+                        # For ZR module, Datapath needes to be re-initlialized on new channel selection
+                        if api.is_coherent_module():
+                           freq = self.port_dict[lport]['laser_freq']
+                           # If user requested frequency is NOT the same as configured on the module
+                           # force datapath re-initialization
+                           if 0 != freq and freq != api.get_laser_config_freq():
```

Comment:
what does it mean when freq == 0?

Reply:
It means 'laser_freq' is missing in PORT table of CONFIG_DB. see get_configured_laser_freq()

Reply:
Understood. It means the no laser_freq is configured in config_db.